File: javascript/orientacaoobjetos/ContaCorrente.js
import { Cliente } from "./Cliente.js";

export class ContaCorrente {
  agencia;
  _cliente;

  set cliente(cliente) {
    if (cliente instanceof Cliente) {
      this._cliente = cliente;
    }
  }

  get cliente(){
      return this._cliente;
  }

  // #saldo =0 https://github.com/tc39/proposal-class-fields#private-fields
  _saldo = 0;

  get saldo(){
      return this._saldo;
  }

  sacar(valor) {
    if (this._saldo >= valor) {
      this._saldo -= valor;
      return valor;
    }
  }

  depositar(valor) {
    if (valor <= 0) {
      return;
    }
    this._saldo += valor;
  }

  transferir(valor, conta) {
    const valorSacado = this.sacar(valor);
    conta.depositar(valorSacado);
  }
}

File: kotlin/PacotesComposicao/src/main/kotlin/com/jotajr/bytebank/modelo/ContaCorrente.kt
package com.jotajr.bytebank.modelo

class ContaCorrente(
    titular: Cliente,
    numero: Int
) : com.jotajr.bytebank.modelo.Conta(
    titular = titular,
    numero = numero
) {
    override fun saca(valor: Double) {
        val valorComTaxa = valor + 0.1
        if(this.saldo >= valorComTaxa){
            this.saldo -= valorComTaxa
        }
    }
}